app/routes/statement.js

```diff
@@ -1,31 +1,52 @@
+const Boom = require('@hapi/boom')
+
 const { getFileStream } = require('../storage')
-const joi = require('joi')
-const boom = require('@hapi/boom')
+const { get, set } = require('../cache')
+
+const streamToBuffer = require('../stream-to-buffer')
+
+const schema = require('./schemas/statement')
 
 module.exports = {
   method: 'GET',
   path: '/statement/{version}/{filename}',
+  options: {
+    validate: {
+      params: schema,
+      failAction: async (request, h, error) => {
+        return Boom.notFound(error)
+      }
+    }
+  },
   handler: async (request, h) => {
     const filename = request.params.filename
 
     try {
-      const statementFile = await getFileStream(filename)
-      return h.response(statementFile.readableStreamBody)
+      const cachedFile = await get(request, filename)
+      if (cachedFile) {
+        console.log(`Cached file found for: ${filename}`)
+
+        return h.response(Buffer.from(cachedFile))
+          .type('application/pdf')
+          .header('Content-Disposition', `attachment;filename=${filename}`)
+          .code(200)
+      }
+
+      console.log('No cached file found, retrieving from storage')
+
+      const fileStream = await getFileStream(filename)
+      const fileBuffer = await streamToBuffer(fileStream.readableStreamBody)
+
+      await set(request, filename, fileBuffer)
+
+      return h.response(fileBuffer)

```

Comment:
I'd consider moving this logic out of the route handler so it just handles the buffered output regardless of whether it came from the cache or not.  Or at least some of it if  not all practical.

Reply:
Agreed -- done